### PR TITLE
Carbon Type Descriptors via ReflectionDescriptor

### PR DESCRIPTION
**Pull request by AlexandruGG**
## Description
This PR adds type descriptors for [Carbon](https://carbon.nesbot.com/) date types Carbon and CarbonImmutable using the ReflectionDescriptor.

Made possible after: https://github.com/briannesbitt/Carbon/pull/2344.

## Features / Changes
- [x] Registers descriptors in `extension.neon`
- [x] Adds test cases in `MyBrokenEntity.php`

### CODE REVIEW

**Comment by AlexandruGG** [comment]
@ondrejmirtes hello! 

Firstly, thank you for this extension and PHPStan in general, it's very useful! 👍 

I'm looking to add type descriptors for [Carbon](https://carbon.nesbot.com/) date types and wanted to check two things first:
1. Is this something that you consider within the scope of this extension? Or is it more something like a custom type that the end-user should add themselves. I thought it would be useful because the library is widely used

2. I'm having trouble running a fresh `composer install` in this project. I tried various things but still could not get it working. Am I missing something?

```
$ composer install
No lock file found. Updating dependencies instead of installing from lock file. Use composer update over composer install if you do not have a lock file.
Loading composer repositories with package information
Updating dependencies
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - phpstan/phpstan-doctrine is present at version 0.12.x-dev and cannot be modified by Composer
    - Conclusion: don't install doctrine/common 3.1.x-dev (conflict analysis result)
    - Conclusion: don't install doctrine/common 3.2.x-dev (conflict analysis result)
    - Conclusion: don't install doctrine/common 3.0.3 (conflict analysis result)
    - Conclusion: don't install doctrine/common 3.1.0 (conflict analysis result)
    - Conclusion: don't install doctrine/common 3.1.1 (conflict analysis result)
    - Conclusion: don't install doctrine/common 3.1.2 (conflict analysis result)
    - Root composer.json requires doctrine/persistence ^1.1 || ^2.0 -> satisfiable by doctrine/persistence[v1.1.0, ..., 1.4.x-dev, 2.0.0, ..., 2.3.x-dev].
    - Root composer.json requires doctrine/mongodb-odm ^1.3 || ^2.1 -> satisfiable by doctrine/mongodb-odm[1.3.0-RC1, ..., 1.3.x-dev, 2.1.0, ..., 2.3.x-dev].
    - doctrine/orm[2.9.1, ..., 2.10.x-dev] require doctrine/common ^3.0.3 -> satisfiable by doctrine/common[3.0.3, ..., 3.2.x-dev].
    - Conclusion: don't install doctrine/common 3.0.x-dev (conflict analysis result)
    - Root composer.json requires doctrine/orm ^2.9.1 -> satisfiable by doctrine/orm[2.9.1, 2.9.x-dev, 2.10.x-dev].```

**Comment by ondrejmirtes** [comment]
Hi, can you link to those Carbon Doctrine types you want to describe? Thanks.

As for your installation problem:

* Make sure you use Composer 2
* Try deleting composer.lock and vendor/.
* Try it with --ignore-platform-reqs

Check out the CI scripts, it's not doing anything special...

**Comment by AlexandruGG** [comment]
> Hi, can you link to those Carbon Doctrine types you want to describe? Thanks.
> 
> As for your installation problem:
> 
> * Make sure you use Composer 2
> * Try deleting composer.lock and vendor/.
> * Try it with --ignore-platform-reqs
> 
> Check out the CI scripts, it's not doing anything special...

Hey, thanks for the fast reply.

I managed to run the installation with `--ignore-platform-reqs` 👍 .

Regarding the types, here they are:
- [CarbonType](https://github.com/briannesbitt/Carbon/blob/master/src/Carbon/Doctrine/CarbonType.php)
- [CarbonImmutableType](https://github.com/briannesbitt/Carbon/blob/master/src/Carbon/Doctrine/CarbonImmutableType.php)

They both use the [CarbonTypeConverter](https://github.com/briannesbitt/Carbon/blob/master/src/Carbon/Doctrine/CarbonTypeConverter.php) to do the conversions from/to database.

**Comment by ondrejmirtes** [comment]
The easiest way forward would be for the library to add proper typehints - the methods convertToPHPValue and convertToDatabaseValue don't have the correct return typehints. And after that registering the ReflectionDescriptor would be sufficient. (https://github.com/phpstan/phpstan-doctrine#reflectiondescriptor).

Alternatively an implementation like https://github.com/phpstan/phpstan-doctrine/blob/master/src/Type/Doctrine/Descriptors/Ramsey/UuidTypeDescriptor.php would also work.

**Comment by AlexandruGG** [comment]
Thanks for the details.

Looks like not even the [Doctrine date types](https://github.com/doctrine/dbal/blob/3.1.x/src/Types/DateImmutableType.php#L24) have those type hints 🤷‍♂️  . 

Also it might be difficult due to the way the code is structured as the [converter](https://github.com/briannesbitt/Carbon/blob/master/src/Carbon/Doctrine/CarbonTypeConverter.php) is used for both Carbon and CarbonImmutable. The return type of `convertToPHPValue` will be the class instance returned by [getCarbonClassName](https://github.com/briannesbitt/Carbon/blob/master/src/Carbon/Doctrine/CarbonTypeConverter.php#L18-L21), which gets overridden in [DateTimeImmutableType](https://github.com/briannesbitt/Carbon/blob/210cc733e1004fc4fb2ab2f9ee8503e54a8f6ead/src/Carbon/Doctrine/DateTimeImmutableType.php#L16-L19). Is it possible to convey this properly via type hints?

The above is mostly why I thought a custom descriptor like [DateTimeImmutableType](https://github.com/phpstan/phpstan-doctrine/blob/master/src/Type/Doctrine/Descriptors/DateTimeImmutableType.php) would be the easiest, unless I'm missing something.

**Comment by ondrejmirtes** [comment]
Oh, in that case we need a custom DoctrineTypeDescriptor implementation here :) So please submit the PR, thanks :)

**Comment by AlexandruGG** [comment]
> Oh, in that case we need a custom DoctrineTypeDescriptor implementation here :) So please submit the PR, thanks :)

Added now, please run the workflow as it's my first time contributing here :)

File: .gitignore [hunk 1]
```diff
@@ -2,3 +2,4 @@
 /vendor
 composer.lock
 .phpunit.result.cache
+.idea
```
Comment (ondrejmirtes) [inline .gitignore:5]: Please don't commit this. It belongs to global .gitignore in your OS.

File: phpunit.xml [hunk 2]
```diff
@@ -10,7 +10,7 @@
          beStrictAboutTodoAnnotatedTests="true"
          failOnRisky="true"
          failOnWarning="true"
-         xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/9.3/phpunit.xsd"
+         xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/9.5/phpunit.xsd"
```
Comment (ondrejmirtes) [inline phpunit.xml:13]: This is an unrelated change - if you're experiencing some error due to this, please open a separate PR with an explanation, thanks.

File: src/Type/Doctrine/Descriptors/Carbon/CarbonImmutableType.php [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+	}
+
+	public function getWritableToPropertyType(): Type
+	{
+		return new ObjectType(CarbonImmutable::class);
+	}
+
+	public function getWritableToDatabaseType(): Type
+	{
+		return new ObjectType(CarbonImmutable::class);
```
Comment (ondrejmirtes) [inline src/Type/Doctrine/Descriptors/Carbon/CarbonImmutableType.php:25]: I'm missing this logic here:

> The return type of convertToPHPValue will be the class instance returned by getCarbonClassName, which gets overridden in here.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: phpunit.xml (hunk 2 above)
Comment (AlexandruGG) [reply phpunit.xml:13]: there's no error but this project uses [phpunit 9.5](https://github.com/phpstan/phpstan-doctrine/blob/master/composer.json#L30). I can leave it for a separate PR since it's unrelated if you think that's best, I just thought it's a minor thing

Comment (ondrejmirtes) [reply phpunit.xml:13]: Yeah, it's minor, but I'm interested in how did you notice this - does PHPUnit print a warning or something?

File: src/Type/Doctrine/Descriptors/Carbon/CarbonImmutableType.php (hunk 3 above)
Comment (AlexandruGG) [reply src/Type/Doctrine/Descriptors/Carbon/CarbonImmutableType.php:25]: you're right. Is this why the tests are failing or did I miss something registering the descriptor?

File: phpunit.xml (hunk 2 above)
Comment (AlexandruGG) [reply phpunit.xml:13]: no warning, I was just curious about the config :)

Comment (ondrejmirtes) [reply phpunit.xml:13]: Yeah, so please remove this change and feel free to submit it separately, but I don't thinkt it's necessary to touch it :)

File: tests/Rules/Doctrine/ORM/EntityColumnRuleTest.php [hunk 4]
```diff
@@ -38,6 +40,12 @@ protected function getRule(): Rule
 		if (!Type::hasType(UuidType::NAME)) {
 			Type::addType(UuidType::NAME, UuidType::class);
 		}
+		if (!Type::hasType(CarbonType::NAME)) {
+			Type::addType(CarbonType::NAME, \Carbon\Doctrine\CarbonType::class);
```
Comment (AlexandruGG) [inline tests/Rules/Doctrine/ORM/EntityColumnRuleTest.php:44]: the `Carbon\Doctrine\CarbonType` doesn't have a NAME constant, so I added it [here](https://github.com/phpstan/phpstan-doctrine/pull/192/files#diff-8c4cceaaf5869a43f944547773744cf916f09124dfb9eaf6fe6290bb83f2e38aR13) instead

File: src/Type/Doctrine/Descriptors/Carbon/CarbonImmutableType.php (hunk 3 above)
Comment (AlexandruGG) [reply src/Type/Doctrine/Descriptors/Carbon/CarbonImmutableType.php:27]: > The return type of convertToPHPValue will be the class instance returned by getCarbonClassName, which gets overridden in here.

Regarding this, I'm not sure it's needed here, everything seems to work as intended now in the tests (added valid case too).

However after playing around a bit with the code in the library itself I reckon I could make it work using generics: https://phpstan.org/r/2a5b7ba7-681b-4a93-b6e7-19ae2fba7e3c. Is the `ReflectionDescriptor` able to figure out the type through that?

Comment (ondrejmirtes) [reply src/Type/Doctrine/Descriptors/Carbon/CarbonImmutableType.php:27]: Try it :)

Comment (AlexandruGG) [reply src/Type/Doctrine/Descriptors/Carbon/CarbonImmutableType.php:27]: update on this: it works 🎉 . I will try to get a PR on nesbot/carbon with those annotations, keeping this open until then. 

After that, can this PR just add these?
```
services:
    -
        factory: PHPStan\Type\Doctrine\Descriptors\ReflectionDescriptor('Carbon\Doctrine\CarbonImmutableType')
        tags: [phpstan.doctrine.typeDescriptor]
    -
        factory: PHPStan\Type\Doctrine\Descriptors\ReflectionDescriptor('Carbon\Doctrine\CarbonType')
        tags: [phpstan.doctrine.typeDescriptor]

Comment (AlexandruGG) [reply src/Type/Doctrine/Descriptors/Carbon/CarbonImmutableType.php:27]: [the PR in Carbon](https://github.com/briannesbitt/Carbon/pull/2344) was merged but I'll wait for the release until re-submitting the ReflectionDescriptor additions

**Comment by AlexandruGG** [comment]
@ondrejmirtes hey 👋 , this is updated and ready for re-review now

**Review by ondrejmirtes — APPROVED** [review]

**Comment by AlexandruGG** [comment]
thank you! 💯
